examples/return_err/lib.rs

```diff
@@ -0,0 +1,74 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[ink::contract]
+mod return_err {
+
+    #[ink(storage)]
+    pub struct ReturnErr {
+        instantiated: bool,
+    }
+
+    /// Example of Error enum
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(::scale_info::TypeInfo))]
+    pub enum Error {
+        Foo,
+    }
+
+    impl Default for ReturnErr {
+        fn default() -> Self {
+            Self { instantiated: true }
+        }
+    }
+
+    /// Type alias for the contract's result type.
+    pub type Result<T> = core::result::Result<T, Error>;
```

Comment:
I don't like this alias. It would make sense if you used this error in more than one occasion. But the error isn't an enum which suggests it is not to be used for more than one function. Yet, you are using the highly contended `Result` identifier for it.

Reply:
We can actually use `Result<Self, ()>` on its own without aliases. If we want to define a type alias for convenience, we need to explicitly provide a full path. I should have made it clearer.

Reply:
And does this work?

```rust
type Foo = Result<MyContract, MyError>;

#[ink(constructor, payable)]
pub fn another_new(fail: bool) -> Foo { ... }
```
or this?
```rust
pub trait Foo {
    type Bar;
}
impl Foo for MyContract {
    type Bar = Result<Self, MyError>;
}
impl MyContract {
    #[ink(constructor, payable)]
    pub fn another_new(fail: bool) -> <MyContract as Foo>::Bar { ... }
}
```
or does this actually not think that the following code represents a fallible ink! constructor?
```rust
type Result = MyContract;

#[ink(constructor, payable)]
pub fn another_new(fail: bool) -> Result { ... }
```
Rust proc. macro hygiene requires proc. macros to work in any semantic contexts regardless of the surrounding context. With a syntactical analysis you can never achieve this and generated code is extremely brittle and easily malicious.

Reply:
Nope the following examples above would not works because the only syntax we accept is `Result<Self, ...>`. Where the `Result` is coming from is not a concern at this stage because we are working with tokens. I can use `is_result_type!()` to also ensure that the `Result` is of the tight type.

Reply:
no, message return types are only checked against *not being `Self`*

Reply:
I mean with regard to `Result`. They also need to know whether you return a result to make sure that in case of an `Err` the proper flags are passed to `seal_return`.  Are they able to detect a `Result` behind an alias like asked for by @Robbepop ?

Reply:
I have not touched `Message`s  but to answer question - yes, they do support type aliasing and the examples above would works for messages

Reply:
Then this leads to the question: Why handle it differently here? The code that handles this seem to exist. Why not reuse it? It seems to be very similar.

Reply:
Because Message's have more permissive syntax checks? Because they can handle more return types by their nature (i.e. anything, but `Self`).

I also checked how messages handle `-> Result<Self, ()>` and while it doesn't compile (as expected). It doesn't give any meaningful error either. I understand that you all included towards type checking, but it doesn't really give us the same guarantees on the code correctness on the syntax level and affect the developer experience IMHO.